### ISSUE TITLE
Constrained systems

### ISSUE THREAD

**Issue by blegat** [JuliaReach/Systems.jl#1]
What do we do to encode constraints on state or input ?
For state constraints, we could do (or `DiscreteConstrainedLinearSystem` I don't know)
```julia
struct ConstrainedDiscreteLinearSystem{T, MT <: AbstractMatrix{T}, ST} <: AbstractDiscreteSystem
    A::MT
    X::ST
end
```
Then for the input we could have
```julia
struct DiscreteLinearControlSystem{T, MT <: AbstractMatrix{T}} <: AbstractDiscreteSystem
    A::MT
    B::MT
end
struct ConstrainedDiscreteLinearControlSystem{T, MT <: AbstractMatrix{T}, ST} <: AbstractDiscreteSystem
    A::MT
    X::ST
    B::MT
    U::ST
end
```
We could also have a type with unconstrained state and constrained input (or constrained state and unconstrained input) but it might be a bit too much, `ST` should have a value equivalent to the full unconstrained space to handle these cases.
We might consider having a different matrix type for `A` and `B` or a different  set type for `X` and `U`.
What do you think ?

**Comment by schillic** [comment]
* I usually prefer `ConstrainedDiscreteLinearSystem` over `DiscreteConstrainedLinearSystem` because it specializes `DiscreteLinearSystem` (for the same reason I would actually prefer `LinearDiscreteSystem` over `DiscreteLinearSystem`).
* What is `ST` in a `DiscreteLinearControlSystem` - it is not used there?
* Agreed, having a type `Unconstrained` or such for `ST` sounds good (maybe even `Void`?). You typically do not create thousands of systems, so a little overhead is fine.
* With the above we could even drop `DiscreteLinearControlSystem` altogether because it is just a `ConstrainedDiscreteLinearControlSystem` where `ST == Unconstrained` (and then could again drop the `Constrained` prefix). But maybe this is going too far and we want to have more specific types...
* Different matrix types: bit of a mess, but it might be good to have the full generality here.

**Comment by mforets** [comment]
let's defer "Different matrix types" as a feature that can be added later if u agree.

> What is ST in a DiscreteLinearControlSystem - it is not used there?

i think it was meant to be 

```
struct DiscreteLinearControlSystem{T, MT <: AbstractMatrix{T}, ST} <: AbstractDiscreteSystem
    A::MT
    B::MT
    U::ST
end
```
to model `x' = Ax + Bu`.

**Comment by mforets** [comment]
like this? (just asking) with CS for ContinuousSystem:

|Name|Model|
|------|------|
|LinearCS|`x' = Ax`|
|ConstrainedLinearCS|`x' = Ax`, `x(t) \in X for all t`|
|LinearControlCS|`x' = Ax + Bu`|
|ConstrainedLinearControlCS|`x' = Ax + Bu`, `x(t) \in X for all t`, `u(t) \in U for all t`|

**Comment by blegat** [comment]
> What is ST in a DiscreteLinearControlSystem - it is not used there?

I have updated the comment.

> With the above we could even drop DiscreteLinearControlSystem altogether because it is just a ConstrainedDiscreteLinearControlSystem where ST == Unconstrained (and then could again drop the Constrained prefix). But maybe this is going too far and we want to have more specific types...

In some fields, it is very natural to have constrained but in other fields it might have a repulsive effet to have to deal with this added complexity.

**Comment by schillic** [comment]
>  In some fields, it is very natural to have constrained but in other fields it might have a repulsive effet to have to deal with this added complexity.

You are right, when somebody searches for the right system in this package, they will be confused.

**Comment by mforets** [comment]
while we are at this, we should consider types for initial value problems = if we want to specify the input at time `t=0` or `k=0`

**Comment by schillic** [comment]
Combinatorial explosion ahead. Maybe it is better to decouple this from the system definition. We could add an outer layer type that contains an `AbstractSystem` and such options. Just a hunch...

**Comment by mforets** [comment]
i've updated the table with more types and initial value problems. indeed, the number of types easily increases. this is not necessarily bad for a library whose aim is to provide definitions for systems, IMO.

on the other hand what do you mean by "outer layer type", more specifically?

for optional field values, [Parameters.jl](https://github.com/mauro3/Parameters.jl), could be useful

**Comment by schillic** [comment]
What I meant is what is described [here](http://www.stochasticlifestyle.com/type-dispatch-design-post-object-oriented-programming-julia/) as "composition".
```julia
type IVP{S<:AbstractSystem}
    system::S
    x0::Type_of_x0
end
```
Otherwise, if we want to go for explicit types for each interesting combination, at the end of the above-mentioned blog post there is a description of [ConcreteAbstractions.jl](https://github.com/tbreloff/ConcreteAbstractions.jl) which can help handle the multiple specializations.

**Comment by mforets** [comment]
thanks for the links, should we give a try to composition then? the initial value problem which extends the `AbstractSystems` seems good to me. `Type_of_x0` should also be a parameter of `IVP`, right?

**Comment by blegat** [comment]
I also like the composition idea for the initial value.

**Comment by mforets** [comment]
ok. branch for constrained continuous systems on the way...

**Comment by blegat** [comment]
Closed by #3
